functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const authURL = oauth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: ['https://www.googleapis.com/auth/gmail.modify'],
+      prompt: 'consent',
+    });
+    res.redirect(authURL);
+  }
+);
+
+const saveTokens = async (values: any) => {
+  const newGmailTokensRef = db.ref(`/gmail/${values.emailKey}`);
+  newGmailTokensRef.set({
+    oauth: {
+      token: values.accessToken,
+      refreshToken: values.refreshToken,
+    },
+    emailAddress: values.fullEmail,
+  });
+};
+
+export const oauth2callback = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { tokens } = await oauth2Client.getToken(req.query.code);
+    oauth2Client.setCredentials(tokens);
+
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const profile = await gm.users.getProfile({ userId: 'me' });
+    const emailAddressKey = profile.data.emailAddress
+      .split('@')[0]
+      .replace('.', '');
+
+    await saveTokens({
+      emailKey: emailAddressKey,
+      fullEmail: profile.data.emailAddress,
+      accessToken: tokens.access_token,
+      refreshToken: tokens.refresh_token,
+    });
+
+    res.redirect(
+      `/Gmail-initWatch?emailAddress=${querystring.escape(profile.data.emailAddress)}`
+    );
+  }
+);
+
+const fetchToken = async (email: string) => {
+  const emailKey = email.split('@')[0].replace('.', '');
+  const queryResults = await db.ref(`/gmail/${emailKey}`).once('value');
+  return queryResults.val();
+};
+
+const storeHistoryIdInDatabase = async (email: string, historyId: any) => {
+  const emailKey = email.split('@')[0].replace('.', '');
+  return await db.ref(`/gmail/${emailKey}`).update({
+    lastSyncHistoryId: historyId,
+  });
+};
+
+export const initWatch = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    if (!req.query.emailAddress) {
+      return res.status(400).send('No emailAddress specified.');
+    }
+    const email = querystring.unescape(req.query.emailAddress);
+    if (!email.includes('@')) {
+      return res.status(400).send('Invalid emailAddress.');
+    }
+
+    // Initiate gmail client
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { oauth, emailAddress } = await fetchToken(email);
+
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
```

Comment:
```suggestion
    const gmail = await Google.google.gmail({
```
I don't like to shorten the names :)